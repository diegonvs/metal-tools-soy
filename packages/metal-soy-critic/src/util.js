const chalk = require('chalk');
const Promise = require('bluebird');

function combineResults(first, second) {
  return toResult(
    first.status && second.status,
    ...first.messages.concat(second.messages)
  );
}

function difference(setA, setB) {
  const difference = new Set();
  for (const elem of setA) {
    if (!setB.has(elem)) {
      difference.add(elem);
    }
  }
  return difference;
}

function toResult(status, ...messages) {
  return {
    messages,
    status
  };
}

function parseTemplateName(rawName) {
  const segments = rawName.split('.');
  const namespace = segments
    .slice(0, segments.length - 1)
    .join('.');

  return {
    name: segments[segments.length - 1],
    namespace: namespace || null
  };
}

function sequence(...tasks) {
  return Promise.reduce(tasks, (res, next) =>
    next().then(value => [...res, value]),
    []
  );
}

function joinErrors(lines) {
  return lines
    .map(line => chalk.red(line))
    .join('\n');
}

module.exports = {
  combineResults,
  difference,
  joinErrors,
  parseTemplateName,
  sequence,
  toResult
};
