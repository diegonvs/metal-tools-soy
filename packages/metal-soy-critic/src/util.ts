import * as chalk from 'chalk';

export interface Result {
  status: boolean;
  messages: Array<string>;
}

export function combineResults(first: Result, second: Result): Result {
  return toResult(
    first.status && second.status,
    ...first.messages.concat(second.messages)
  );
}

export function toResult(status: boolean, ...messages: Array<string>): Result {
  return {
    messages,
    status
  };
}

export interface TemplateName {
  name: string,
  namespace: string | null
}

export function difference<T>(setA: Set<T>, setB: Set<T>): Set<T> {
  const difference = new Set();
  for (const elem of setA) {
    if (!setB.has(elem)) {
      difference.add(elem);
    }
  }
  return difference;
}

export function joinErrors(lines: Array<string>): string {
  return lines
    .map(line => chalk.red(line))
    .join('\n');
}

export function isSorted(items: Array<string>): boolean {
  const sortedItems = items
    .concat()
    .sort((a, b) => a.localeCompare(b));

  return sortedItems.join() === items.join();
}
