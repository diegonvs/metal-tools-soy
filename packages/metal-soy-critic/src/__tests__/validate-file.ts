import {getFixturePath}  from '../../test/util';
import validateFile from '../validate-file';

describe('Validate file', () => {
  function validate(fileName) {
    return validateFile(getFixturePath(fileName));
  }

  test('should pass', () => validate('Test.soy').then(result => {
      expect(result.status).toBe(true);
  }));

  test('should fail; missing params', () => validate('MissingParams.soy').then(result => {
      expect(result.status).toBe(false);
      expect(result.messages).toMatchSnapshot();
  }));

  test('should fail; missing internal', () => validate('MissingInternal.soy').then(result => {
      expect(result.status).toBe(false);
      expect(result.messages).toMatchSnapshot();
  }));

  test('should fail; missing import', () => validate('MissingCallImport.soy').then(result => {
      expect(result.status).toBe(false);
      expect(result.messages).toMatchSnapshot();
  }));

  test('should fail; missing required', () => validate('MissingRequired.soy').then(result => {
      expect(result.status).toBe(false);
      expect(result.messages).toMatchSnapshot();
  }));

  test('should fail; invalid namespace', () => validate('InvalidNamespace.soy').then(result => {
      expect(result.status).toBe(false);
      expect(result.messages).toMatchSnapshot();
  }));

  test('should fail; missing render', () => validate('MissingRender.soy').then(result => {
      expect(result.status).toBe(false);
      expect(result.messages).toMatchSnapshot();
  }));

  test('should fail; invalid javascript', () => validate('InvalidJS.soy').then(result => {
      expect(result.status).toBe(false);
      expect(result.messages).toMatchSnapshot();
  }));
});
