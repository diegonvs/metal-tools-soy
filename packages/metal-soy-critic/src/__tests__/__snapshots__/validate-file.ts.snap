// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Validate file should fail; invalid namespace 1`] = `
Array [
  "Failed to parse soy template",
]
`;

exports[`Validate file should fail; missing import 1`] = `
Array [
  "It looks like the following component calls are missing an [33mimport[39m:

[31mMyExternalComponent[39m",
]
`;

exports[`Validate file should fail; missing internal 1`] = `
Array [
  "Based on their name, these attributes should have the [33m.internal()[39m config added:

[31mcount_[39m",
]
`;

exports[`Validate file should fail; missing params 1`] = `
Array [
  "The following params should be included in [33mSTATE[39m:

[31moptionalInfo[39m",
]
`;

exports[`Validate file should fail; missing render 1`] = `
Array [
  "Cannot find a [33m.render[39m template, do you have a typo?

[31mrander[39m",
]
`;

exports[`Validate file should fail; missing required 1`] = `
Array [
  "These attributes are [33mrequired[39m in your Soy Template but not in your Component:

[31mtitle[39m",
  "These attributes are [33m.required()[39m in your Component but not in your Template:

[31mdescription[39m",
]
`;
